lib/Echidna/RPC.hs

```diff
@@ -0,0 +1,146 @@
+import EVM.Exec (exec, vmForEthrunCreation)
+import EVM.Types (Addr, W256)
+
+import qualified Control.Monad.Fail as M (MonadFail(..))
+import qualified Control.Monad.State.Strict as S (state)
+import qualified Data.ByteString.Base16 as BS16 (decode)
+import qualified Data.Text as T (Text, drop)
+import qualified Data.Vector as V (fromList)
+
+import Echidna.Exec
+import Echidna.Transaction
+
+-- | During initialization we can either call a function or create an account or contract
+data Etheno = AccountCreated Addr
+            | ContractCreated Addr Addr Integer Integer ByteString W256
+            | FunctionCall Addr Addr Integer Integer ByteString W256
+  deriving (Eq, Show)
+
+instance FromJSON Etheno where
+  -- parseJSON = genericParseJSON $ defaultOptions {fieldLabelModifier = tail, omitNothingFields = True}
+  parseJSON = withObject "Etheno" $ \v -> do
+    (ev :: String) <- v .: "event"
+    case ev of
+         "AccountCreated"  -> AccountCreated  <$> v .: "address"
+         "ContractCreated" -> ContractCreated <$> v .: "from"
+                                              <*> v .: "contract_address"
+                                              <*> (read <$> (v .: "gas_used"))
+                                              <*> (read <$> (v .: "gas_price"))
+                                              <*> (decode <$> (v .: "data"))
+                                              <*> v .: "value"
+         "FunctionCall"    -> FunctionCall    <$> v .: "from"
+                                              <*> v .: "to"
+                                              <*> (read <$> (v .: "gas_used"))
+                                              <*> (read <$> (v .: "gas_price"))
+                                              <*> (decode <$> (v .: "data"))
+                                              <*> v .: "value"
+         _ -> M.fail "event should be one of \"AccountCreated\", \"ContractCreated\", or \"FunctionCall\""
+    where decode :: T.Text -> ByteString
+          decode = decodeHex . encodeUtf8 . T.drop 2
+          decodeHex = fst . BS16.decode
+
+
+-- | Handler for parsing errors
+data EthenoException = EthenoException String
+
+instance Show EthenoException where
+    show (EthenoException e) = "Error parsing Etheno initialization file: " ++ e
+
+instance Exception EthenoException
+
+
+-- | Main function: takes a filepath where the initialization sequence lives and returns
+-- | the initialized VM along with a list of Addr's to put in GenConf
+loadEthenoBatch :: (MonadThrow m, MonadIO m, Has TxConf y, MonadReader y m) => [T.Text] -> FilePath -> m (VM, [Addr])
+loadEthenoBatch ts fp = do
+  bs <- liftIO $ eitherDecodeFileStrict fp
+
+  case bs of
+       (Left e) -> throwM $ EthenoException e
+       (Right ethenoInit) -> do
+         -- Separate out account creation txns to use later for config
+         let (accounts, txs) = partition (\case { AccountCreated{} -> True; _ -> False; }) ethenoInit
```

Comment:
since we are interested in executing the `txs` that probably doesn't quite work, and the partial fmap is guaranteed to be total because `partition` so there's no real reason for an additional `catMaybes`.

Reply:
I think I'd still prefer to avoid checking in partial functions where it's avoidable. What about `Data.Either.partitionEithers . fmap (\case AccountCreated a -> Left a; b -> Right b)`?

Reply:
that would work

Reply:
no longer fails silently